csrc/multidevice/utils.cpp

```diff
@@ -464,6 +471,27 @@ void insertShardedAxisReordering(Fusion* fusion) {
   }
 }
 
+void setShardedAllocationDomain(Fusion* fusion) {
+  for (auto expr : fusion->exprs()) {
+    if (isResharding(expr)) {
+      for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        for (auto c : tv->getContiguity()) {
+          if (c.has_value()) {
+            NVF_CHECK(
```

Comment:
I'm confused. I thought setShardedAllocationDomain ought to make inputs of resharding exprs contiguous rather than expect them to be contiguous. Am I missing something?

Reply:
^^^ that doesn't sound right. You cannot *change* contiguity / stride order on inputs.

IIUC, the code here validates that the input entry is contiguous and then later explicitly sets allocation domain on each if they are implicit.

One part I'm not totally sure is, `Resharding expression input must be contiguous `. Should this check also apply to TensorViews which we are not specifying allocation domain?

Reply:
ProcessGroup only accept [contiguous tensors](https://github.com/pytorch/pytorch/blob/a00a99e801c14d2b09533e09a9c20a1af473d9ab/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2168) (because nccl and ucc only deal with contiguous buffers, passed as void pointers). So for now it is reasonable to only support contiguous tensors. Later, we could add support for non-contiguous tensors, but to do that we'll have no choice but to make as many process group call as there a contiguous components.

So imo this "assert" makes sense for this pr and we could remove it later by implementing what I described above

@wujingyue this line
https://github.com/NVIDIA/Fuser/blob/690134dbf2403546978631653ea0f5bfabb352d3/csrc/multidevice/utils.cpp#L393-L395

is a "trick" to allow non-oputermost resharding while avoiding non-contiguous buffers, by reordering the axis to place the sharded axis at outermost position (therefore the actual buffer is contiguous in memory)